python/python/bystro/proteomics/annotation_interface.py

```diff
@@ -0,0 +1,188 @@
+    """Flatten an arbitrarily nested list."""
+    if not isinstance(xs, list):
+        return [xs]
+    return sum([_flatten(x) for x in xs], [])
+
+
+def _get_samples_genes_dosages_from_hit(hit: dict[str, Any]) -> pd.DataFrame:
+    """Given a document hit, return a dataframe of samples, genes and dosages."""
+    source = hit["_source"]
+    gene_names = _flatten(source["refSeq"]["name2"])
+    # homozygotes, heterozygotes may not be present in response, so
+    # represent them as empty lists if not.
+    heterozygotes = _flatten(source.get("heterozygotes", []))
+    homozygotes = _flatten(source.get("homozygotes", []))
+    missing_genos = _flatten(source.get("missingGenos", []))
+    rows = []
+    for gene_name in gene_names:
+        for heterozygote in heterozygotes:
+            rows.append(
+                {"sample_id": heterozygote, "gene_name": gene_name, "dosage": HETEROZYGOTE_DOSAGE}
+            )
+        for homozygote in homozygotes:
+            rows.append({"sample_id": homozygote, "gene_name": gene_name, "dosage": HOMOZYGOTE_DOSAGE})
+        for missing_geno in missing_genos:
+            rows.append(
+                {"sample_id": missing_geno, "gene_name": gene_name, "dosage": MISSING_GENO_DOSAGE}
+            )
+
+    return pd.DataFrame(rows)
+
+
+def _execute_query(
+    client: OpenSearch,
+    query_args: dict,
+) -> pd.DataFrame:
+    """Process OpenSearch query and return results."""
+    resp = client.search(**query_args)
+    return _process_response(resp)
+
+
+def _process_response(resp: dict[str, Any]) -> pd.DataFrame:
+    """Postprocess query response from opensearch client."""
+    num_hits = len(resp["hits"]["hits"])
+    total_value = resp["hits"]["total"]["value"]
+    if num_hits != total_value:
+        err_msg = f"Number of hits: {num_hits} didn't equal total value: {total_value}. This is a bug."
+        raise ValueError(err_msg)
+
+    samples_genes_dosages_df = pd.concat(
+        [_get_samples_genes_dosages_from_hit(hit) for hit in resp["hits"]["hits"]]
+    )
+    # we may have multiple variants per gene in the results, so we
+    # need to drop duplicates here.
+    return samples_genes_dosages_df.drop_duplicates()
+
+
+def _get_num_slices(
+    client: OpenSearch,
+    index_name: str,
+    query: dict[str, Any],
+) -> int:
+    """Count number of hits for the index."""
+    get_num_slices_query = query["body"].copy()
+    get_num_slices_query.pop("sort", None)
+    get_num_slices_query.pop("track_total_hits", None)
+
+    response = client.count(body=get_num_slices_query, index=index_name)
+
+    n_docs: int = response["count"]
+    if n_docs < 1:
+        err_msg = (
+            f"Expected at least one document in `response['count']`, got response: {response} instead."
+        )
+        raise RuntimeError(err_msg)
+
+    num_slices_necessary = math.ceil(n_docs / OPENSEARCH_QUERY_CONFIG.max_query_size)
+    num_slices_planned = min(num_slices_necessary, OPENSEARCH_QUERY_CONFIG.max_slices)
```

Comment:
Why do we want `num_slices_planned`? If we have more than 10,000 * 1024 records, we want to return all of them

Reply:
per discussion off-line, holding off on this comment until we determine why this implementation was chosen for search/save/handler.py

Reply:
Remind myself of the API: https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#scroll-search-results

The sliced scroll is distinct from regular scroll in that there are multiple scroll contexts, which evenly divide (if possible) the total hits. Each scroll context is paginated independently, from the first document accessible to that scroll context, to the last, and pagination happens in "size" records.

So this use, and the one in the save handler, are fine.